File: demos/julia/timer_demo.jl
#!/usr/bin/env julia
#=
Timer Demo — Verifies the Timer API works correctly.
=#
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "bindings", "julia", "src"))

using Insight
using Insight: Timer, timer_start, timer_stop, timer_elapsed_ms, timer_destroy

# CPU Timer
let
    t = Timer(0, 0)
    timer_start(t)
    sleep(0.005)
    timer_stop(t)
    ms = timer_elapsed_ms(t)
    println(string("CPU timer: ", round(ms, digits=3), " ms"))
    if ms < 0.0 || ms > 100.0
        error("CPU timer out of range: $ms")
    end
    timer_destroy(t)
end

# GPU Timer (if available)
if is_device_available(DeviceKindGPU)
    load_backend("cuda")
    t = Timer(1, 0)
    timer_start(t)
    a = ones([256, 256], DTypeF32, GPUPlace(0))
    b = full([256, 256], 2.0, DTypeF32, GPUPlace(0))
    c = add(a, b)
    timer_stop(t)
    ms = timer_elapsed_ms(t)
    println(string("GPU timer: ", round(ms, digits=3), " ms"))
    timer_destroy(t)
else
    println("GPU timer: SKIPPED (GPU not available)")
end

println("OK")

File: demos/cpp/timer_demo.cpp
// demos/cpp/timer_demo.cpp
/// Timer Demo — Verifies the Timer API works correctly.
#include "insight/core/profiler.h"
#include "insight/insight.h"
#include "insight/ops/creation.h"
#include "insight/ops/elementwise.h"
#include <chrono>
#include <iostream>
#include <thread>

int main() {
  ins::init();

  // CPU Timer
  {
    ins::Timer timer(ins::CPUPlace());
    timer.start();
    std::this_thread::sleep_for(std::chrono::milliseconds(5));
    timer.stop();
    float ms = timer.elapsed_ms();
    std::cout << "CPU timer: " << ms << " ms" << std::endl;
    if (ms < 0.0f || ms > 100.0f) {
      std::cerr << "ERROR: CPU timer out of range: " << ms << " ms"
                << std::endl;
      return 1;
    }
  }

  // GPU Timer (if available)
  if (ins::is_device_available(ins::DeviceKind::GPU)) {
    ins::Timer timer(ins::GPUPlace(0));
    timer.start();
    ins::Array a = ins::ones({512, 512}, ins::DType::F32, ins::GPUPlace(0));
    ins::Array b =
        ins::full({512, 512}, 2.0f, ins::DType::F32, ins::GPUPlace(0));
    ins::Array c = ins::add(a, b);
    timer.stop();
    float ms = timer.elapsed_ms();
    std::cout << "GPU timer: " << ms << " ms" << std::endl;
    if (ms < 0.0f) {
      std::cerr << "ERROR: GPU timer returned negative: " << ms << " ms"
                << std::endl;
      return 1;
    }
  } else {
    std::cout << "GPU timer: SKIPPED (GPU not available)" << std::endl;
  }

  std::cout << "OK" << std::endl;
  return 0;
}
